More codegen cleanups

nnethercote opened the pull request:
Some additional cleanups I found while looking closely at this code, following up from #112827.

r= @oli-obk

nnethercote committed:
Inline and remove `WorkItem::start_profiling` and `execute_work_item`.

They both match on a `WorkItem`. It's simpler to do it all in one place.

oli-obk commented on compiler/rustc_interface/src/queries.rs:
```diff
@@ -323,7 +317,7 @@ impl<'tcx> Queries<'tcx> {
             prepare_outputs,
             crate_hash,
-            ongoing_codegen,
+            ongoing_codegen: ongoing_codegen,
```
This diff is not necessary

nnethercote replied:
:+1:

oli-obk reviewed (changes requested):
r=me with nit

oli-obk commented:
r? @oli-obk

nnethercote committed:
Remove Queries::ongoing_codegen.

There's no need to store it in `Queries`. We can just use a local
variable, because it's always used shortly after it's produced.

The commit also removes the `tcx.analysis()` call in `ongoing_codegen`,
because it's easy to ensure that's done beforehand.

All this makes the dataflow within `run_compiler` easier to follow, at
the cost of making one test slightly more verbose, which I think is a
good tradeoff.

nnethercote commented:
The first commit caused the crash, and I don't understand why. I have removed it, I'll try doing it in a separate PR. The other two commits can be merged here.

@bors r=oli-obk rollup

bors commented:
:pushpin: Commit 1da13489247ed5b0b6694b2aabb0184f16b50afd has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk 1da13489247ed5b0b6694b2aabb0184f16b50afd -->
<!-- homu: {"type":"Approved","sha":"1da13489247ed5b0b6694b2aabb0184f16b50afd","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

nnethercote commented:
> The first commit caused the crash, and I don't understand why. I have removed it, I'll try doing it in a separate PR.

It was because I changed the type of `__rustc_codegen_backend` but didn't also change its callers. (This isn't caught by the type checker because of the code structure.) I found a simpler way to do it that doesn't change the type in #112913.
